### DISCUSSION TITLE
Identifying photos that are in iCloud, not local library

### DISCUSSION THREAD

**Discussion by Rott-Apple** [RhetTbull/osxphotos#1082] [General]
I apologize if this question has been asked before. My use case is this: Photos in the iCloud Photos Library with 'Optimize Mac Storage' turned on - so full sized photos are in iCloud; smaller copies in the Photos Library on the Mac (I may have downloaded some full sized photos for editing or exporting, so the local library will have a mix of full and smaller sized photos).

I don't know if I have enough space to download all the photos whose full sized versions _aren't_ in the local library. How to find the total size of such photos?

Also, can I use OSXPhotos to find out the total size of a selection of photos (e.g. a few albums) whose full sized versions _aren't_ in the local library?

Many thanks!

**Comment by RhetTbull** [comment]
You can get a list of all photos that are in iCloud but not on the local machine using the following command:

`osxphotos query --not-shared --missing --incloud > missing.csv`

This will create a file `missing.csv` that is in comma separated value (CSV) form which can be opened in Microsoft Excel, Google Sheets, etc. One of the columns will be original file size and you could sum those. This is only an approximation as it wouldn't include the size of the edited file if the file had been edited.  A photo is reported as missing with the `--missing` flag if Photos reports the original as not being on the local disk. In some cases if a photo has been edited, the original might be missing but the edited version is on disk or vice versa. It's difficult to know the exact size without writing a custom script.

One proxy that would be close is to examine the size reported by iCloud for your photos then looking at the size of the local library. The difference between these two is close to the amount of disk space you would need.

Edit: The `--not-shared` flag eliminates shared photos (from shared albums) from the query as these will always be reported as missing since Photos does not keep originals of shared images, only a lower resolution jpeg.

**Reply by RhetTbull** [reply]
> Also, can I use OSXPhotos to find out the total size of a selection of photos (e.g. a few albums) whose full sized versions aren't in the local library?

You could use the same command as above but add `--selected` to query only photos currently selected or for specific albums, use `--album "Album name 1" --album "Album name 2"` with the query above,

**Reply by Rott-Apple** [reply]
Thank you, that's extremely helpful. I'll give both commands a try.

As for the proxy you suggested, I find that the size of the local library is actually a bit larger than the size shown by iCloud. I guess this is because as long as the Mac has space, full sized versions of all photos are stored in iCloud as well as on the Mac. Only when the Mac is out of space, smaller-sized versions will be generated locally and full-sized versions left only in iCloud. At present the local library is larger because it includes the databases, thumbnails and other overheads. Does this sound right?

**Reply by RhetTbull** [reply]
> At present the local library is larger because it includes the databases, thumbnails and other overheads. Does this sound right?

That make sense though I imagine the iCloud library also stores some of that (you can browse photos at iCloud.com). Since Apple doesn't document the library structure or how iCloud syncing for Photos works it's hard to know for certain. I checked my personal library and iCloud reports 119 GB and the local library is 153 GB. The odd thing is my originals on the Mac are 123 GB (not including edited photos) which is less that iCloud is reporting.

**Reply by rajscode** [reply]
APFS / Clones / block allocation could all be playing into what is being "shown" for local storage usage vs how they are storing and calculating space usage in iCloud (which I would guess is not using APFS volumes *LOL*.  Prob on a linux system with zfs or a raw volume where the datastore is managing the space allocation ( just guesses here on my part 
 ( ¯\_(ツ)_/¯ )

Try a search on your favorite search engine with the following: 
apfs copy reported size site:eclecticlight.co

Lots of interesting info and details at [eclecticlight.co](url)

See this one for example: https://eclecticlight.co/2023/04/23/last-week-on-my-mac-when-the-numbers-dont-add-up/

FYI - My full copy on my local APFS volume shows the library @ 1.58 TB
On iCloud it shows 1.42 TB space.

**Reply by rajscode** [reply]
> FYI - My full copy on my local APFS volume shows the library @ 1.58 TB On iCloud it shows 1.42 TB space.

One other thing re: my particular full library on the local APFS volume.  It is 100 % been populated from the iCloud version.  I've never added a photo / video or any other item to that particular library via any mechanism (photo imports from disk, usb drives or even attaching the iPhone or iPad, etc..).  The only "import" that would have been done to that local library would be due to an osxphotos export --download-missing --use-photokit command ... 

Am pretty confident re: the sizing re: space used locally vs cloud isn't accurate

**Comment by Rott-Apple** [comment]
Does ZINTERNALRESOURCE.ZDATALENGTH being smaller than ZADDITIONALASSETATTRIBUTES.ZORIGINALFILESIZE indicate the full-sized versions are in iCloud library and optimized ones in local library? ZINTERNALRESOURCE has multiple records per ZASSET or ZADDITIONALASSETATTRIBUTES record, presumably one for every thumbnail stored in Originals and the various Resources sub-folders. Filtering out ZFILEID -1 items removes most of these but still leaves some duplicates. If these remaining duplicates could be filtered out, might this be a robust way to figure out what will be downloaded and thus the space required?

The APFS angle is interesting; thanks for the links.

**Comment by trader0007** [comment]
what about the other way around ?  i appear to have 12 more photos locally than in the cloud (and photos says everything is sync-ed up). can i find which ones are in local that are not in cloud ?

**Reply by RhetTbull** [reply]
You could try `osxphotos query --not-incloud` for photos that aren't synced yet or `osxphotos query --not-cloudasset` for photos which are not cloud assets at all. This can happen if you have referenced files  (stored outside the Photos library) in a library that is also synced to iCloud. The referenced files won't be synced.
